File: internal/tokenomics/models/token.go
package models

import (
	"math/big"
	"time"
)

// TokenConfig represents the core ECHO token specification
type TokenConfig struct {
	Name        string
	Symbol      string
	TotalSupply *big.Int // 1 billion * 10^8
	Decimals    int
	HardCapped  bool
}

// NewTokenConfig creates the ECHO token configuration
func NewTokenConfig() *TokenConfig {
	totalSupply := new(big.Int)
	totalSupply.SetString("100000000000000000", 10) // 1 billion * 10^8

	return &TokenConfig{
		Name:        "ECHO",
		Symbol:      "ECHO",
		TotalSupply: totalSupply,
		Decimals:    8,
		HardCapped:  true,
	}
}

// AllocationBreakdown defines WO-214 genesis pool distribution (1B ECHO fixed supply).
type AllocationBreakdown struct {
	CommunityRewards *big.Int // 400M — 10-year emission account
	Treasury         *big.Int // 220M — 3-of-5 multi-sig
	Founders         *big.Int // 180M — TokenLock positions
	FutureTeam       *big.Int // 100M — multi-sig controlled
	Ecosystem        *big.Int // 100M — governance-controlled
}

// NewAllocationBreakdown creates the WO-214 genesis allocation breakdown.
func NewAllocationBreakdown() *AllocationBreakdown {
	echo := func(millions int64) *big.Int {
		// millions * 1e6 ECHO * 1e8 datum
		return new(big.Int).Mul(big.NewInt(millions), big.NewInt(100_000_000_000000))
	}
	return &AllocationBreakdown{
		CommunityRewards: echo(400),
		Treasury:         echo(220),
		Founders:         echo(180),
		FutureTeam:       echo(100),
		Ecosystem:        echo(100),
	}
}

// TokenBalance tracks user balance and vesting
type TokenBalance struct {
	Address          string
	AvailableBalance *big.Int
	VestingSchedule  *VestingSchedule
}

// VestingSchedule defines time-locked token releases
type VestingSchedule struct {
	TotalAmount   *big.Int
	ReleasedAt    time.Time
	CliffMonths   int
	VestMonths    int
	ReleasedSoFar *big.Int
}

// CalculateReleasable calculates currently releasable amount
func (vs *VestingSchedule) CalculateReleasable() *big.Int {
	elapsed := time.Since(vs.ReleasedAt)
	monthsElapsed := int(elapsed.Hours() / 730)

	if monthsElapsed < vs.CliffMonths {
		return big.NewInt(0)
	}

	if monthsElapsed >= vs.CliffMonths+vs.VestMonths {
		return new(big.Int).Sub(vs.TotalAmount, vs.ReleasedSoFar)
	}

	vestingMonths := monthsElapsed - vs.CliffMonths
	monthlyRelease := new(big.Int).Div(vs.TotalAmount, big.NewInt(int64(vs.VestMonths)))
	releasable := new(big.Int).Mul(monthlyRelease, big.NewInt(int64(vestingMonths)))

	releasable.Sub(releasable, vs.ReleasedSoFar)
	return releasable
}

File: ios/Echo/Sources/Models/Token.swift
import Foundation

/// Token model representing ECHO token specifications
public struct TokenConfig {
    public let name: String = "ECHO"
    public let symbol: String = "ECHO"
    public let totalSupply: Decimal = 1_000_000_000
    public let decimals: Int = 8
    public let hardCapped: Bool = true
    
    public init() {}
}

/// Allocation breakdown for token distribution (per ECHO Tokenomics v2.0)
public struct AllocationBreakdown {
    public let communityRewards: Decimal  // 40% - 400M (auto-scaling model, no daily caps)
    public let treasury: Decimal          // 22% - 220M
    public let founders: Decimal          // 18% - 180M (CEO 100M + 4 co-founders 20M each)
    public let futureTeam: Decimal        // 10% - 100M
    public let validatorRewards: Decimal  // 10% - 100M

    public init() {
        let total = Decimal(1_000_000_000)
        self.communityRewards = total * Decimal(string: "0.40")!
        self.treasury = total * Decimal(string: "0.22")!
        self.founders = total * Decimal(string: "0.18")!
        self.futureTeam = total * Decimal(string: "0.10")!
        self.validatorRewards = total * Decimal(string: "0.10")!
    }

    /// Calculate total allocation to verify correctness
    public func totalAllocation() -> Decimal {
        return communityRewards + treasury + founders + futureTeam + validatorRewards
    }

    /// Founder allocation detail (per Tokenomics v2.0)
    public struct FounderAllocation {
        public static let ceoAllocation: Decimal = 100_000_000      // 10% - CEO
        public static let coFounderAllocation: Decimal = 20_000_000  // 2% each - 4 co-founders
        public static let multiSigThreshold = 3                      // 3-of-5 founder multi-sig
        public static let multiSigTotal = 5
        public static let cliffMonths = 12
        public static let vestingMonths = 36                         // 1/36th monthly after cliff
        public static let withdrawCooldownDays = 14
    }
}

/// Token balance tracking
public struct TokenBalance {
    public let address: String
    public var availableBalance: Decimal
    public var vestingSchedule: VestingSchedule?
    
    public init(address: String, availableBalance: Decimal = 0) {
        self.address = address
        self.availableBalance = availableBalance
    }
}

/// Vesting schedule for locked tokens
public struct VestingSchedule {
    public let totalAmount: Decimal
    public let releasedAt: Date
    public let cliffMonths: Int
    public let vestMonths: Int
    public private(set) var releasedSoFar: Decimal = 0
    
    public init(totalAmount: Decimal, releasedAt: Date, cliffMonths: Int, vestMonths: Int) {
        self.totalAmount = totalAmount
        self.releasedAt = releasedAt
        self.cliffMonths = cliffMonths
        self.vestMonths = vestMonths
    }
    
    /// Calculate releasable amount
    public func calculateReleasable() -> Decimal {
        let calendar = Calendar.current
        let now = Date()
        
        let releaseComponents = calendar.dateComponents([.year, .month], from: releasedAt, to: now)
        
        let monthsElapsed = (releaseComponents.year ?? 0) * 12 + (releaseComponents.month ?? 0)
        
        // Before cliff period
        if monthsElapsed < cliffMonths {
            return 0
        }
        
        // After full vesting period
        if monthsElapsed >= cliffMonths + vestMonths {
            return totalAmount - releasedSoFar
        }
        
        // During vesting
        let vestingMonths = monthsElapsed - cliffMonths
        let monthlyRelease = totalAmount / Decimal(vestMonths)
        let totalReleasable = monthlyRelease * Decimal(vestingMonths)
        
        return totalReleasable - releasedSoFar
    }
}
